src/utils/vcs.rs

```diff
@@ -668,3 +798,443 @@ fn test_url_normalization() {
+}
+
+#[cfg(test)]
+fn git_commit_test(dir: &Path, file_path: &str, content: &[u8], commit_message: &str) {
+    let path = dir.join(file_path);
+    let mut file = File::create(path).expect("Failed to execute.");
+    file.write_all(content).expect("Failed to execute.");
+
+    let mut add = Command::new("git")
+        .args(&["add", "."])
+        .current_dir(dir)
+        .spawn()
+        .expect("Failed to execute `git add .`");
+
+    add.wait().expect("Failed to wait on git add.");
+
+    let mut commit = Command::new("git")
+        .args(&[
+            "commit",
+            "-m",
+            commit_message,
+            "--author",
+            "John Doe <john.doe@example.com>",
+        ])
+        .current_dir(dir)
+        .spawn()
+        .expect("Failed to execute `git commit -m {message}`.");
+
+    commit.wait().expect("Failed to wait on git commit.");
+}
+
+#[test]
+fn test_get_commits_from_git() {
+    let dir = tempdir().expect("Failed to generate temp dir.");
+    test_initialize(dir.path());
+    git_commit_test(
+        dir.path(),
+        "foo.js",
+        b"console.log(\"Hello, world!\");",
+        "\"initial commit\"",
+    );
+
+    git_commit_test(
+        dir.path(),
+        "foo2.js",
+        b"console.log(\"Hello, world! Part 2\");",
+        "\"second commit\"",
+    );
+
+    let repo = git2::Repository::open(dir.path()).expect("Failed");
+    let commits = get_commits_from_git(&repo, "").expect("Failed");
+
+    assert_debug_snapshot!(commits.0.iter().map(|c| {
+        (c.author().name().unwrap().to_owned(), c.author().email().unwrap().to_owned(),c.summary())
+    }).collect::<Vec<_>>(), @r###"
+    [
+        (
+            "John Doe",
+            "john.doe@example.com",
+            Some(
+                "\"second commit\"",
+            ),
+        ),
+        (
+            "John Doe",
+            "john.doe@example.com",
+            Some(
+                "\"initial commit\"",
+            ),
+        ),
+    ]
+    "###);
+}
+
+#[test]
+fn test_generate_patch_set_base() {
+    let dir = tempdir().expect("Failed to generate temp dir.");
+    test_initialize(dir.path());
+    git_commit_test(
+        dir.path(),
+        "foo.js",
+        b"console.log(\"Hello, world!\");",
+        "\"initial commit\"",
+    );
+
+    git_commit_test(
+        dir.path(),
+        "foo2.js",
+        b"console.log(\"Hello, world! Part 2\");",
+        "\"second commit\"",
+    );
+
+    git_commit_test(
+        dir.path(),
+        "foo2.js",
+        b"console.log(\"Hello, world! Part 3\");",
+        "\"third commit\"",
+    );
+
+    let repo = git2::Repository::open(dir.path()).expect("Failed");
+    let commits = get_commits_from_git(&repo, "").expect("Failed");
+    println!("Commits: {}", commits.0.len());
+    let patch_set =
+        generate_patch_set(&repo, commits.0, commits.1, "example/test-repo").expect("Failed");
+
+    assert_yaml_snapshot!(patch_set, {
```

Comment:
I think we can store these in separate files. You can ask Armin about how to do that.